src/EditorFeatures/Core/Implementation/BraceMatching/BraceHighlightingViewTaggerProvider.cs

```diff
@@ -19,44 +22,64 @@ namespace Microsoft.CodeAnalysis.Editor.Implementation.BraceMatching
-            : base(new AggregateAsynchronousOperationListener(asyncListeners, FeatureAttribute.BraceHighlighting),
-                   notificationService)
+                : base(new AggregateAsynchronousOperationListener(asyncListeners, FeatureAttribute.BraceHighlighting), notificationService)
         {
             _braceMatcherService = braceMatcherService;
         }
 
-        protected override bool RemoveTagsThatIntersectEdits => true;
-
-        protected override SpanTrackingMode SpanTrackingMode => SpanTrackingMode.EdgeExclusive;
+        public override ITaggerEventSource CreateEventSource(ITextView textView, ITextBuffer subjectBuffer)
+        {
+            return TaggerEventSources.Compose(
+                TaggerEventSources.OnTextChanged(subjectBuffer, TaggerDelay.NearImmediate),
+                TaggerEventSources.OnCaretPositionChanged(textView, subjectBuffer, TaggerDelay.NearImmediate));
+        }
 
-        protected override IEnumerable<Option<bool>> TagSourceOptions
+        public override Task ProduceTagsAsync(TaggerContext<BraceHighlightTag> context, DocumentSnapshotSpan documentSnapshotSpan, int? caretPosition)
         {
-            get
+            var document = documentSnapshotSpan.Document;
+            if (!caretPosition.HasValue || document == null)
             {
-                yield return InternalFeatureOnOffOptions.BraceMatching;
+                return SpecializedTasks.EmptyTask;
             }
+
+            return ProduceTagsAsync(context, document, documentSnapshotSpan.SnapshotSpan.Snapshot, caretPosition.Value);
         }
 
-        protected override ITaggerEventSource CreateEventSource(ITextView textView, ITextBuffer subjectBuffer)
+        internal async Task ProduceTagsAsync(TaggerContext<BraceHighlightTag> context, Document document, ITextSnapshot snapshot, int position)
         {
-            return TaggerEventSources.Compose(
-                TaggerEventSources.OnTextChanged(subjectBuffer, TaggerDelay.NearImmediate),
-                TaggerEventSources.OnCaretPositionChanged(textView, subjectBuffer, TaggerDelay.NearImmediate));
+            using (Logger.LogBlock(FunctionId.Tagger_BraceHighlighting_TagProducer_ProduceTags, context.CancellationToken))
+            {
+                await ProduceTagsForBracesAsync(context, document, snapshot, position, rightBrace: false).ConfigureAwait(false);
+                await ProduceTagsForBracesAsync(context, document, snapshot, position - 1, rightBrace: true).ConfigureAwait(false);
+            }
         }
 
-        protected override ITagProducer<BraceHighlightTag> CreateTagProducer()
+        private async Task ProduceTagsForBracesAsync(TaggerContext<BraceHighlightTag> context, Document document, ITextSnapshot snapshot, int position, bool rightBrace)
         {
-            return new BraceHighlightingTagProducer(_braceMatcherService);
+            if (position >= 0 && position < snapshot.Length)
```

Comment:
Position can be the end of the file if you have a close squiggle there -- we highlight if you're on the outside of the squiggle. Shouldn't this be position <= snapshot.Length?

A comment at least that explains why we're passing in potentially bad locations would be appreciated. This has a code smell of "I didn't understand how snapshots work so I just hacked around it".